statsmodels/formula/_manager.py

```diff
@@ -0,0 +1,723 @@
+        if self._engine == "patsy":
+            return patsy.ModelDesc.from_formula(formula)
+        else:
+            return formulaic.Formula(formula)
+
+    def get_term_names(self, spec_or_frame):
+        if not isinstance(spec_or_frame, self.model_spec_type):
+            _spec = self.get_model_spec(spec_or_frame)
+        else:
+            _spec = spec_or_frame
+        if self._engine == "patsy":
+            return list(_spec.term_names)
+        else:
+            return [str(term) for term in _spec.terms]
+
+    def get_column_names(self, spec_or_frame):
+        if not isinstance(spec_or_frame, self.model_spec_type):
+            spec = self.get_model_spec(spec_or_frame)
+        else:
+            spec = spec_or_frame
+        return list(spec.column_names)
+
+    @property
+    def model_spec_type(self):
+        if self._engine == "patsy":
+            return patsy.design_info.DesignInfo
+        else:
+            return formulaic.model_spec.ModelSpec
+
+    def get_term_name_slices(self, spec_or_frame):
+        if not isinstance(spec_or_frame, self.model_spec_type):
+            spec = self.get_model_spec(spec_or_frame)
+        else:
+            spec = spec_or_frame
+        if self._engine == "patsy":
+            return spec.term_name_slices
+        else:
+            return spec.term_slices
+
+    def get_model_spec(self, frame, optional=False):
+        if self._engine == "patsy":
+            if optional and not hasattr(frame, "design_info"):
+                return None
+            return frame.design_info
+        else:
+            if optional and not hasattr(frame, "model_spec"):
+                return None
+            return frame.model_spec
+
+    def get_slice(self, model_spec, term):
+        if self._engine == "patsy":
+            return model_spec.slice(term)
+        else:
+            return model_spec.get_slice(term)
+
+    def get_term_name(self, term):
+        if self._engine == "patsy":
+            return term.name()
+        else:
+            return str(term)
+
+    def get_description(self, spec_or_frame):
+        if not isinstance(spec_or_frame, self.model_spec_type):
+            _spec = self.get_model_spec(spec_or_frame)
+        else:
+            _spec = spec_or_frame
+        if self._engine == "patsy":
+            return _spec.describe()
+        else:
+            return str(_spec.formula)
+
+    def get_factor_categories(self, factor, model_spec):
+        if self._engine == "patsy":
+            return model_spec.factor_infos[factor].categories
+        else:
+            return tuple(model_spec.encoder_state[factor][1]["categories"])
+
+    def get_contrast_matrix(self, term, factor, model_spec):
+        if self._engine == "patsy":
+            return model_spec.term_codings[term][0].contrast_matrices[factor].matrix
+        else:
+            cat = self.get_factor_categories(factor, model_spec)
+            reduced_rank = True
+            for ts in model_spec.structure:
```

Comment:
This won't work in general; since a term may consist of the product of than one categorical variable. I'll add a helper method for this. I toyed with the notion of adding a helper method for this, but looking at this code I should have just done it. I'll add a method `model_spec.get_factor_contrasts_for_term(....)`; the the final should be: `model_spec.get_factor_contrasts_for_term(term).get_coding_matrix()`

Reply:
The problem is that I don't know what to do with interaction terms.
I have several open issues to support both categorical-categorical and categorical-continuous interactions, e.g.

For t_test_pairwise, predict and marginal/predicitive effects, one requirement for categorical-categorical interaction would be to get all cell means, e.g. partial exog for the columns with possibly setting other columns to prespecified values.
One option I thought about is to automatically create artificial datasets, but I never tried to figure out how to do this with patsy, e.g. #8439  #7071 

There are several post-estimation results that need more information about the structure of terms in the design matrix. But those are currently not supported.
Currently only anova_lm and t_test_pairwise use term information, and t_test_pairwise is restricted to a single factor.